File: ds/fundamentals/stack/stack.py
# Author: Francis Osei Annin
# Date: 14/04/23
# Description: A stack data structure

class Stack(object):
    """A stack is a data structure, that follows the last in, first out principle."""

    def __init__(self):
        """initialise stack with empty element list"""
        self.elements = []

    def push(self, key):
        """add key to the top of the stack"""
        self.elements.append(key)

    def pop(self):
        """remove first element of stack"""
        if not self.is_empty():
            return self.elements.pop()
        else:
            return Exception("Can't pop from empty stack")

    def peek(self):
        """return the first element on stack"""
        if not self.is_empty():
            return self.elements[-1]
        else:
            return Exception("Can't peek on empty stack")

    def size(self):
        """return the size of the stack"""
        return len(self.elements)

    def is_empty(self):
        """return boolean if size is empty or not"""
        return self.elements == 0

    def sweet_pop(self):
        """reduce the top element of the stack"""
        if not self.is_empty():
            del self.elements[-1]
        else:
            return Exception("Underflow Error, can't peek on empty stack")

File: ds/fundamentals/stack/go/stack.go
// author: francis osei annin
// date: 29/08/2024
package main

type Stack struct {
	container []any
}

// Push adds a new element on top of the stack
func (p *Stack) Push(element any) {
	p.container = append(p.container, element)
}

// Pop removes the element on top of the stack
func (p *Stack) Pop() any {
	p.container = p.container[0 : len(p.container)-1]
	return p.container[len(p.container)-1]
}

// Peek show first element on top of stack, without removing it
func (p *Stack) Peek() any {
	return p.container[len(p.container)-1]
}

// IsEmpty checks if a stack is element or not
func (p *Stack) IsEmpty() bool {
	if p.Size() != 0 {
		return false
	}
	return true
}

// Size returns the number of element in a stack
func (p *Stack) Size() any {
	return len(p.container)
}
